qutebrowser/browser/hints.py

```diff
@@ -132,7 +144,41 @@ def _move_to_elem(self):
     def cleanup(self):
         """Clean up this element and hide it."""
         self.hide()
-        self.deleteLater()
+        self.setParent(None)
+        self._context.tab.contents_size_changed.disconnect(self._move_to_elem)
+        HintLabel._object_pool.append(self)
+
+        HintLabel._active_object_count -= 1
+        if HintLabel._active_object_count == 0:
+            active_object_peak_count = HintLabel._active_object_peak_count
+
+            # Shrink the object pool if it is not used for the last 5 peaks
+            # and it has more than 500 elements
+            if len(HintLabel._object_pool) > 500:
+                del HintLabel._object_pool[max(active_object_peak_count):]
```

Comment:
If we have 501 elements, and we hint 1 element 5 times, everything will be deleted. On the other hand, if we have 499 elements and hint 1 element 5 times, nothing will be deleted. Maybe we shouldn't delete more than 500 in this case (or whatever we decide the max is).

Reply:
@user202729 if you get some free time could you look at this? I think this is really helpful overall, we just need to iron out the logic for mainatining the pool so there's no weird edge cases. 

FWIW, I took a look at the memory usage after hinting on reddit, and I couldn't find any very large offenders in hinting (the largest being about 30kb), so it might be easiest to just set a large limit on the hint pool (which most people will never reach). I'm fine with continuing down the dynamic limit route though, whatever you think is better :).